File: platforms/ios/Dezel/Dezel/View/Display/Display.swift
/**
 * @class Display
 * @since 0.7.0
 */
public class Display {

	//--------------------------------------------------------------------------
	// MARK: Properties
	//--------------------------------------------------------------------------

	/**
	 * @property window
	 * @since 0.7.0
	 */
	public var window: DisplayNode? {
		willSet {
			DisplaySetWindow(self.handle, newValue?.handle)
		}
	}

	/**
	 * @property scale
	 * @since 0.7.0
	 */
	public var scale: Double = 1 {
		willSet {
			DisplaySetScale(self.handle, newValue)
		}
	}

	/**
	 * @property viewportWidth
	 * @since 0.7.0
	 */
	public var viewportWidth: Double = 0 {
		willSet {
			DisplaySetViewportWidth(self.handle, newValue)
		}
	}

	/**
	 * @property viewportHeight
	 * @since 0.7.0
	 */
	public var viewportHeight: Double = 0 {
		willSet {
			DisplaySetViewportHeight(self.handle, newValue)
		}
	}

	/**
	 * @property stylesheet
	 * @since 0.7.0
	 */
	public var stylesheet: Stylesheet? {
		willSet {
			DisplaySetStylesheet(self.handle, newValue?.handle)
		}
	}

	/**
	 * @property invalid
	 * @since 0.7.0
	 */
	public var invalid: Bool {
		return DisplayIsInvalid(self.handle)
	}

	/**
	 * @property resolving
	 * @since 0.7.0
	 */
	public var resolving: Bool {
		return DisplayIsResolving(self.handle)
	}

	/**
	 * @property handle
	 * @since 0.7.0
	 * @hidden
	 */
	private(set) public var handle: DisplayRef

	/**
	 * @property prepareCallbacks
	 * @since 0.7.0
	 * @hidden
	 */
	private var prepareCallbacks: [(() -> Void)] = []

	/**
	 * @property resolveCallbacks
	 * @since 0.7.0
	 * @hidden
	 */
	private var resolveCallbacks: [(() -> Void)] = []

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.7.0
	 */
	internal init() {
		self.handle = DisplayCreate()
		DisplaySetPrepareCallback(self.handle, displayPrepareCallback)
		DisplaySetResolveCallback(self.handle, displayResolveCallback)
		DisplaySetData(self.handle, UnsafeMutableRawPointer(value: self))
	}

	/**
	 * @destructor
	 * @since 0.7.0
	 */
	deinit {
		DisplayGetData(self.handle).release()
		DisplaySetData(self.handle, nil)
		DisplayDelete(self.handle)
	}

	/**
	 * @method resolve
	 * @since 0.7.0
	 */
	public func resolve() {
		DisplayResolve(self.handle)
	}

	/**
	 * @method registerPrepareCallback
	 * @since 0.7.0
	 */
	public func registerPrepareCallback(_ callback: @escaping () -> Void) {
		self.prepareCallbacks.append(callback)
	}

	/**
	 * @method registerResolveCallback
	 * @since 0.7.0
	 */
	public func registerResolveCallback(_ callback: @escaping () -> Void) {
		self.resolveCallbacks.append(callback)
	}

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * @method didPrepare
	 * @since 0.7.0
	 * @hidden
	 */
	internal func didPrepare() {
		self.dispatchPrepareCallback()
	}

	/**
	 * @method didResolve
	 * @since 0.7.0
	 * @hidden
	 */
	internal func didResolve() {
		self.dispatchResolveCallback()
	}

	/**
	 * @method dispatchLayoutBeganEvent
	 * @since 0.7.0
	 * @hidden
	 */
	internal func dispatchPrepareCallback() {
		self.prepareCallbacks.forEach { $0() }
		self.prepareCallbacks.removeAll()
	}

	/**
	 * @method dispatchLayoutEndedEvent
	 * @since 0.7.0
	 * @hidden
	 */
	internal func dispatchResolveCallback() {
		self.resolveCallbacks.forEach { $0() }
		self.resolveCallbacks.removeAll()
	}
}

/**
 * @function displayPrepareCallback
 * @since 0.7.0
 * @hidden
 */
private let displayPrepareCallback: @convention(c) (DisplayRef?) -> Void = { ptr in
	if let display = DisplayGetData(ptr).value as? Display {
		display.didPrepare()
	}
}

/**
 * @function displayResolveCallback
 * @since 0.7.0
 * @hidden
 */
private let displayResolveCallback: @convention(c) (DisplayRef?) -> Void = { ptr in
	if let display = DisplayGetData(ptr).value as? Display {
		display.didResolve()
	}
}

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/view/display/Display.kt
package ca.logaritm.dezel.view.display

import ca.logaritm.dezel.view.display.external.DisplayExternal

/**
 * @class Display
 * @since 0.7.0
 */
public class Display {

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * @property window
	 * @since 0.7.0
	 */
	public var window: DisplayNode? = null
		set(value) = DisplayExternal.setWindow(this.handle, value?.handle ?: 0)

	/**
	 * @property scale
	 * @since 0.7.0
	 */
	public var scale: Double = 1.0
		set(value) = DisplayExternal.setScale(this.handle, value)

	/**
	 * @property viewportWidth
	 * @since 0.7.0
	 */
	public var viewportWidth: Double = 0.0
		set(value) = DisplayExternal.setViewportWidth(this.handle, value)

	/**
	 * @property viewportHeight
	 * @since 0.7.0
	 */
	public var viewportHeight: Double = 0.0
		set(value) = DisplayExternal.setViewportHeight(this.handle, value)

	/**
	 * @property viewportHeight
	 * @since 0.7.0
	 */
	public var stylesheet: Stylesheet? = null
		set(value) = DisplayExternal.setStylesheet(this.handle, value?.handle ?: 0)

	/**
	 * @property invalid
	 * @since 0.7.0
	 */
	public val invalid: Boolean
		get() = DisplayExternal.isInvalid(this.handle)

	/**
	 * @property resolving
	 * @since 0.7.0
	 */
	public val resolving: Boolean
		get() = DisplayExternal.isResolving(this.handle)

	/**
	 * @property handle
	 * @since 0.7.0
	 * @hidden
	 */
	public var handle: Long
		private set

	/**
	 * @property prepareCallback
	 * @since 0.7.0
	 * @hidden
	 */
	private var prepareCallback: MutableList<() -> Unit> = mutableListOf()

	/**
	 * @property resolveCallback
	 * @since 0.7.0
	 * @hidden
	 */
	private var resolveCallback: MutableList<() -> Unit> = mutableListOf()

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.7.0
	 */
	init {
		this.handle = DisplayExternal.create(this)
		DisplayReference.register(this)
	}

	/**
	 * @method registerPrepareCallback
	 * @since 0.7.0
	 */
	public fun registerPrepareCallback(callback: (() -> Unit)) {
		this.prepareCallback.add(callback)
	}

	/**
	 * @method registerResolveCallback
	 * @since 0.7.0
	 */
	public fun registerResolveCallback(callback: (() -> Unit)) {
		this.resolveCallback.add(callback)
	}

	//--------------------------------------------------------------------------
	// Private API
	//--------------------------------------------------------------------------

	/**
	 * @method onPrepare
	 * @since 0.7.0
	 * @hidden
	 */
	@Suppress("unused")
	private fun onPrepare() {
		this.dispatchPrepareCallback()
	}

	/**
	 * @method onResolve
	 * @since 0.7.0
	 * @hidden
	 */
	@Suppress("unused")
	private fun onResolve() {
		this.dispatchResolveCallback()
	}

	/**
	 * @method dispatchPrepareCallback
	 * @since 0.7.0
	 * @hidden
	 */
	private fun dispatchPrepareCallback() {
		this.prepareCallback.forEach { it() }
		this.prepareCallback.clear()
	}

	/**
	 * @method dispatchResolveCallback
	 * @since 0.7.0
	 * @hidden
	 */
	private fun dispatchResolveCallback() {
		this.resolveCallback.forEach { it() }
		this.resolveCallback.clear()
	}
}